signature/src/signatures/sigstore.rs

```diff
@@ -0,0 +1,167 @@
+        if config.default_docker.is_some() {
+            if merged_config.default_docker.is_some() {
+                return Err(anyhow!(
+                    "Error parsing sigstore config: \"default-docker\" defined repeatedly."
+                ));
+            }
+            merged_config.default_docker = config.default_docker;
+        }
+
+        // An image namespace is not allowed appear in two different configuration files.
+        if let Some(docker_config_map) = config.docker {
+            for (ns_name, ns_config) in docker_config_map.iter() {
+                if merged_config.is_config_entry_exit(ns_name) {
+                    return Err(anyhow!(format!(
+                        "Error parsing sigstore config: {} defined repeatedly.",
+                        &ns_name
+                    )));
+                }
+
+                merged_config.insert_a_docker_config_entry(ns_name, ns_config);
+            }
+        }
+    }
+
+    Ok(merged_config)
+}
+
+// returns an URL string (sigstore base url) configured in sigstore config for the image.
+pub fn get_sigstore_base_url(image_ref: &Reference) -> Result<String> {
+    let sigstore_config = load_sigstore_config(REGISTRY_CONFIG_DIR)?;
+
+    if let Some(docker_config_map) = sigstore_config.docker {
+        // Look for a full match.
+        let image_id = image_ref.whole();
+        // Look for a match of the possible parent namespaces.
+        let mut image_ns = image::get_image_namespaces(image_ref);
+
+        image_ns.insert(0, image_id);
+        for ns in image_ns.iter() {
+            if let Some(base_url) = docker_config_map.get(ns) {
+                return Ok(base_url.sigstore.to_string());
+            }
+        }
+    }
+
+    // Look for a default location
+    if let Some(default_config) = sigstore_config.default_docker {
+        return Ok(default_config.sigstore);
+    }
+
+    Err(anyhow!(
+        "There is no config info for this image in sigstore config files."
+    ))
+}
+
+pub fn get_sigs_from_specific_sigstore(sigstore: String) -> Result<Vec<Vec<u8>>> {
+    let mut res: Vec<Vec<u8>> = Vec::new();
+    let sigstore_uri = url::Url::parse(&sigstore[..])
+        .map_err(|e| anyhow!("Failed to parse sigstore_uri: {:?}", e))?;
+
+    // FIXME: Now only support get signatures from local files.
+    // only the uri with "file" scheme can be parsed.
+    //
+    // issue: https://github.com/confidential-containers/image-rs/issues/9
+    if sigstore_uri.scheme() == "file" {
+        let sigstore_dir_path = sigstore_uri.path().to_string();
+        for entry in fs::read_dir(sigstore_dir_path)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_dir() {
+                continue;
+            }
+            let path_str = path
+                .to_str()
+                .ok_or(anyhow!("Unknown error: path parsed failed."))?;
+            let sig = fs::read(path_str)?;
```

Comment:
Here (and in the policy module) the signature is represented as a `Vec<u8>` but I suggest you to use a higher-level abstraction. For example:
```rust
struct Signature {
  content: Vec<u8>,
  path: URL,
  sigstore: Sigstore, 
}

struct Sigstore {
  namespace: RegistryNamespace,
  path: URL,
  type: "docker"
}
```

Because you might need to find the reference to the signature path (from where you read it) in case of, for example, throw an error when the signature is invalid.

Reply:
This is a good idea, but I think this optimization can be implemented through a new commit after the current version is merged, because it will lead to a lot of code changes.